packages/material-ui/src/ButtonBase/ButtonBase.js

```diff
@@ -89,20 +77,19 @@ class ButtonBase extends React.Component {
 
   handleContextMenu = createRippleHandler(this, 'ContextMenu', 'stop');
 
-  handleBlur = createRippleHandler(this, 'Blur', 'stop', () => {
-    clearTimeout(this.focusVisibleTimeout);
+  handleBlur = createRippleHandler(this, 'Blur', 'stop', event => {
+    handleBlur(event);
     if (this.state.focusVisible) {
       this.setState({ focusVisible: false });
     }
   });
 
   componentDidMount() {
-    listenForFocusKeys(ownerWindow(this.getButtonNode()));
-
+    prepareFocusVisible();
     if (this.props.action) {
       this.props.action({
         focusVisible: () => {
-          this.setState({ focusVisible: true });
+          this.setState({ focusVisible: true }); // TODO needed?
```

Comment:
Don't comment in code. Use issue trackers.

Reply:
We need to rerender to apply the class name?

Reply:
I assume the "needed?" question refers to the `focusVisible` action and not the `setState` within it. It doesn't look like this is used within Material-UI and I wouldn't think there should be a good reason for someone else to call it. It is documented though, so it would be a breaking change to remove it. I think it would be worthwhile to remove it though, since the new focus-visible logic should be more reliable than the old logic and if people are using this they probably shouldn't going forward. The case mentioned in #9712 when this was added probably required this because of using a hot key to set focus other than one of the keys that was being listened to in the old implementation.

Reply:
The `focusVisible` state is used by the ripple logic. For instance, it will "pulsate". Now, if we can change the logic to purely rely on CSS, it would be a valuable simplification. It might be out of reach for the core problem we want to solve.

Reply:
For the action, we had people complaining that calling .focus() doesn't trigger the ripple state. I see two possible motivation with the specification. 1. Our previous focus visible logic wasn't working as expected, the user was calling .focus() after a keyboard event, the pulsating state should have been present. 2. The user wants to ignore the best accessibility practices, he wants to show a ripple even if no keyboard action was triggered by the users. Let me find the history.

Reply:
It's unclear, it could be a mix of the two: https://github.com/mui-org/material-ui/issues/3008.

Reply:
I would leave it for now. It should technically work to just call `focus` on the element if there was a keyboard event before. This could be abused to set focus-visible that doesn't match the UA heuristic but let's leave this untouched for now. We can discuss removing it later.